Scheduler priority

iomartin opened the pull request:
One more buildrequest priority PR (following #7008, #7010, #7012 and #7016).

The goal of this PR is to allow schedulers to set a default priority for buildrequests it creates. It supports either an integer or a callable, to which we pass the builderNames and the codebases + changes related to that buildrequest. The function must return an integer.

The integer path is pretty straightforward, but for the callable option, should I be passing something else? The `changesByCodebase` is a dictionary whose keys are codebase names and the values is a list of changes (as dictionaries).

Feedback welcome on the design.

I would also like to add more robust tests, the ones I added only check that we set the scheduler.priority to the right value, it doesn't check that buildrequests were actually created with the right priority. But I got a bit confused with the framework/fake changes etc, so help is appreciated.

## Contributor Checklist:

* [X] I have updated the unit tests
* [X] I have created a file in the `newsfragments` directory (and read the `README.txt` in that directory)
* [X] I have updated the appropriate documentation

iomartin commented:
Fixed the CI error

p12tic commented on master/docs/manual/customization.rst:
```diff
@@ -254,6 +254,40 @@ The ``nextBuild`` function is passed as parameter to :class:`BuilderConfig`:
+
+.. code-block:: python
+
+   def schedulerPriority(builderNames, changesByCodebase):
```
snake_case

p12tic reviewed (approved):
Looks great, had just one minor comment.

p12tic commented on master/buildbot/test/util/scheduler.py:
```diff
@@ -224,7 +224,8 @@ def sourceStampKey(sourceStamp):
     def fake_addBuildsetForChanges(self, waited_for=False, reason='', external_idstring=None,
-                                   changeids=None, builderNames=None, properties=None, **kw):
+                                   changeids=None, builderNames=None, properties=None,
+                                   priority=None, **kw):
```
The purpose of this function is to register how `addBuildsetForChanges` would have been called if it was real. Thus add `priority` to addBuildsetForChanges list of dicts below.

p12tic commented:
For a more real test you can copy one of basic scheduler test. For example, if you copied `SingleBranchScheduler.test_gotChange_no_treeStableTimer_important`, setup custom priority (either int or callback), then you should be able to check `self.addBuildsetCalls` for priority values.

p12tic commented:
The design looks sensible. I had more comments about the tests.

iomartin commented:
The `SingleBranchScheduler` tests do not have `self.addBuildsetCalls`, I'm looking into `test_timed_Nightly.py` which uses it

iomartin commented:
Added a test with custom priority that checks the buildrequest is created with that priority.
I didn't add one with the ballback because it would need a more significant change in the framework since the tests use the fake_addBuildset* functions.

I also added a couple few more commits to cleanup and reorganize a few things that I'd changed in my previous PRs

p12tic commented:
> The SingleBranchScheduler tests do not have self.addBuildsetCalls

It's correct that SingleBranchScheduler tests don't currently use `self.addBuildsetCalls`, however they have  `attachScheduler(..., overrideBuildsetMethods=True)` as part of their setup which provides setup for `self.addBuildsetCalls`. Have you found otherwise? I think priority tests aren't related to nightly scheduler, but to base functionality, so they should go to `test_basic.py` file.

p12tic commented:
Thanks!
